src/webgpu/shader/validation/expression/call/builtin/dot.spec.ts

```diff
@@ -0,0 +1,189 @@
+const builtin = 'dot';
+export const description = `
+Validation tests for the ${builtin}() builtin.
+`;
+
+import { makeTestGroup } from '../../../../../../common/framework/test_group.js';
+import { keysOf, objectsToRecord } from '../../../../../../common/util/data_tables.js';
+import { kValue } from '../../../../../util/constants.js';
+import {
+  Type,
+  kConvertableToFloatVectors,
+  scalarTypeOf,
+  ScalarType,
+} from '../../../../../util/conversion.js';
+import { QuantizeFunc, quantizeToF16, quantizeToF32 } from '../../../../../util/math.js';
+import { ShaderValidationTest } from '../../../shader_validation_test.js';
+
+import {
+  fullRangeForType,
+  kConstantAndOverrideStages,
+  stageSupportsType,
+  validateConstOrOverrideBuiltinEval,
+} from './const_override_validation.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+const kValidArgumentTypes = objectsToRecord(kConvertableToFloatVectors);
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function quantizeFunctionForScalarType(type: ScalarType): QuantizeFunc<any> {
+  switch (type) {
+    case Type.f32:
+      return quantizeToF32;
+    case Type.f16:
+      return quantizeToF16;
+    case Type.abstractInt:
+      return BigInt;
+    default:
+      return Number;
+  }
+}
+
+export interface CheckFunc {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  (num: any): boolean;
+}
+
+function checkFunctionForScalarType(type: ScalarType): CheckFunc {
+  switch (type) {
+    case Type.abstractInt:
+      return (num: bigint) => !kValue.i64.isOOB(num);
+    default:
+      return (num: number) => Number.isFinite(num);
+  }
+}
+
+interface StepArgs {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  [name: string]: any;
```

Comment:
The `checkEquation` stuff is cool, but I am wondering if it will have use beyond `dot` tests. If not, it might be easier to understand if you just did the explicit checks without the abstractions. It would also remove the need for `any` here.

Reply:
It could be used for most of the builtin tests I've done so far (normalize, cross, fma, etc) And if I removed the abstractions I'd still have to account for the differences between BigInt and number, which would probably lead to nearly as much supporting code. 

That said, I'm not against approaching it from a different direction as long as we can find a pattern that can keep the equations being tested reasonably clear and separate from the supporting cast/quantize/check code.